### PR TITLE
feat(resilience): add React error boundaries (app + sections + plugins)

### LINKED ISSUE #268: Add a global React ErrorBoundary (app + plugins + panels)
## Problem

\`apps/geolibre-desktop/src/main.tsx\` renders \`<App/>\` with no React error boundary anywhere in the tree. A single render-time throw — a malformed layer style, a crashing plugin component, or a bad \`.geolibre.json\` — white-screens the entire app, leaving only a \`console.error\`.

This is especially risky because GeoLibre loads arbitrary user data and runs **trusted-but-third-party plugin code** (the desktop CSP permits \`blob:\` script execution by design).

## Proposal

- Add a **top-level ErrorBoundary** around \`<App/>\` that renders a recovery panel and surfaces the already-captured diagnostics (\`installDiagnosticsCapture\` in \`main.tsx\`), with a "reload" / "report" action.
- Wrap each **plugin control** and each **panel** in its own boundary so one bad plugin or panel degrades gracefully (shows an inline error + teardown) instead of taking down the whole map. This composes with the existing plugin teardown/unregister logic.

## Acceptance criteria

- A thrown error in a plugin control or panel does not blank the app.
- The top-level boundary shows a recoverable UI and includes diagnostics context.
- A unit/integration test simulates a child throw and asserts the fallback renders.

## Notes

Lowest-cost, highest-impact resilience improvement available right now. No external dependency strictly required, though \`react-error-boundary\` is an option.

### PR DESCRIPTION
**Pull request by giswqs**
Closes #268.

## Problem

`apps/geolibre-desktop/src/main.tsx` rendered `<App/>` with **no error boundary anywhere in the tree**. A single render-time throw — a malformed layer style, a crashing panel, or a bad `.geolibre.json` — white-screened the entire app, leaving only a `console.error`. This is especially risky because GeoLibre loads arbitrary user data and runs trusted-but-third-party plugin code.

## Changes

- **Reusable primitive** — `ErrorBoundary` added to `@geolibre/ui`: presentation-agnostic (caller supplies the fallback render prop), with `onError` reporting and `resetKeys` for automatic recovery when inputs change.
- **App-level boundary** — `AppErrorBoundary` wraps `<App/>` in `main.tsx` and shows a full-screen recovery panel (Try again / Reload app) instead of a blank page.
- **Section boundaries** — `SectionErrorBoundary` wraps each shell region in `DesktopShell` (toolbar, layer panel, map, style panel, attribute table, status bar). A crash in one region shows a compact inline notice with a Retry button while the rest of the app keeps working.
- **Plugin hardening** — plugin controls are imperative MapLibre code, so their throws escape React boundaries. The `toggle`/`setMapControlPosition` paths in `usePlugins.ts` now `try/catch` and report failures so one bad plugin can't break the handler.
- **Diagnostics integration** — every boundary-caught error and plugin failure is recorded as a `runtime` diagnostic, so it's visible and copyable in the existing Diagnostics panel rather than console-only.

## Tests

- New `tests/error-boundary.test.ts`: covers the boundary's derived-state capture and the diagnostics reporting helper.
- Full suite green: `npm run test:frontend` (371 tests), `npm run lint` (0 errors), `npm run build` (typecheck + bundle), and `pre-commit run --files` all pass.

## Notes

React error boundaries only catch errors thrown during rendering/lifecycle of React descendants — not event handlers, async code, or imperative non-React code. The imperative paths (plugin controls, MapLibre internals) remain covered by the existing global `window.error` diagnostics capture, now complemented by the targeted plugin `try/catch`.

## Summary by CodeRabbit

* **New Features**
  * App-wide and per-section error boundaries so individual UI areas (toolbar, map, panels, tables, status) fail and recover independently.
  * Full-screen app fallback offering "Try again" and "Reload app"; compact inline fallbacks with "Retry" for sections.
  * Diagnostics capture for component and plugin errors; error boundaries used at startup.
  * UI library now exposes a reusable ErrorBoundary component and reset-key utility.

* **Tests**
  * Added unit tests for error boundary behavior, reset-key logic, and diagnostics reporting.

### CODE REVIEW

File: apps/geolibre-desktop/src/hooks/usePlugins.ts [hunk 1]
```diff
@@ -158,7 +175,14 @@ export function usePluginRegistry() {
...
-      manager.toggle(id, appApi);
+      // Plugin controls are imperative MapLibre code, so a throw here escapes
+      // React's error boundaries. Contain it so one bad plugin can't break the
+      // toggle handler — surface it in diagnostics instead.
+      try {
+        manager.toggle(id, appApi);
+      } catch (error) {
+        reportPluginError(id, "toggle", error);
+      }
       persistProjectPluginState(before);
```
[the remark this answers is not in the document]
Comment (giswqs) [reply apps/geolibre-desktop/src/hooks/usePlugins.ts:192]: Done in 4f66fb4 — both `toggle` and `setMapControlPosition` now `return` in the catch block, so a half-applied failure is no longer persisted.

File: apps/geolibre-desktop/src/main.tsx [hunk 2]
```diff
@@ -37,10 +37,15 @@ installDiagnosticsCapture();
 installStaleChunkReload();
 
 void import("./App")
-  .then(({ default: App }) => {
+  .then(async ({ default: App }) => {
+    const { AppErrorBoundary } = await import(
+      "./components/common/error-boundaries"
+    );
```
[the remark this answers is not in the document]
Comment (giswqs) [reply apps/geolibre-desktop/src/main.tsx]: Done in 4f66fb4 — the App and error-boundary chunks are now fetched in parallel via `Promise.all`, removing the import waterfall.

File: packages/ui/src/components/error-boundary.tsx [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    return this.props.children;
+  }
+}
+
+function resetKeysChanged(
+  prev: readonly unknown[] | undefined,
+  next: readonly unknown[] | undefined,
+): boolean {
+  if (prev === next) return false;
+  if (!prev || !next || prev.length !== next.length) return true;
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/ui/src/components/error-boundary.tsx:85]: Done in 4f66fb4 — `resetKeysChanged` now treats absent and empty key lists identically, so `undefined`↔`[]` transitions no longer trigger a reset.

File: packages/ui/src/components/error-boundary.tsx [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+  ErrorBoundaryState
+> {
+  state: ErrorBoundaryState = { error: null };
+
+  static getDerivedStateFromError(error: Error): ErrorBoundaryState {
+    return { error };
+  }
+
+  componentDidCatch(error: Error, info: React.ErrorInfo): void {
+    this.props.onError?.(error, info);
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/ui/src/components/error-boundary.tsx:55]: Done in 4f66fb4 — added a doc note on `reportBoundaryError` explaining the expected dev-mode double entry (React re-throws boundary errors as a window.error event in dev; production records once). Placed it in the desktop diagnostics wrapper rather than the generic UI primitive, which has no diagnostics knowledge.

File: tests/error-boundary.test.ts [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+  ({ clearDiagnostics, getDiagnosticsSnapshot } = await import(
+    "../apps/geolibre-desktop/src/lib/diagnostics"
+  ));
+});
+
+describe("ErrorBoundary derived state", () => {
+  it("captures the thrown error into state", () => {
+    const error = new Error("boom");
+    assert.deepEqual(ErrorBoundary.getDerivedStateFromError(error), { error });
+  });
```
[the remark this answers is not in the document]
Comment (giswqs) [reply tests/error-boundary.test.ts:49]: Done in 4f66fb4 — exported `resetKeysChanged` and added tests covering the undefined-vs-empty edge case plus unchanged/changed/count-change cases.

File: apps/geolibre-desktop/src/components/common/error-boundaries.tsx [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+        >
+          <AlertTriangle className="h-10 w-10 text-destructive" />
+          <div className="space-y-1">
+            <h1 className="text-lg font-semibold">Something went wrong</h1>
+            <p className="max-w-md text-sm text-muted-foreground">
+              GeoLibre hit an unexpected error and could not continue. Your work
+              may be unsaved — try recovering before reloading.
+            </p>
+            <p className="max-w-md break-words font-mono text-xs text-muted-foreground/80">
+              {error.message}
```
[the remark this answers is not in the document]
Comment (giswqs) [reply apps/geolibre-desktop/src/components/common/error-boundaries.tsx]: Done in a25fa00 — the app fallback now renders `error.message || String(error)`, so non-Error throws no longer produce a blank panel.

File: apps/geolibre-desktop/src/components/common/error-boundaries.tsx [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+ */
+export function SectionErrorBoundary({
+  label,
+  children,
+  className,
+  resetKeys,
+}: {
+  label: string;
+  children: ReactNode;
+  className?: string;
```
[the remark this answers is not in the document]
Comment (giswqs) [reply apps/geolibre-desktop/src/components/common/error-boundaries.tsx]: Done in a25fa00 — renamed the prop to `fallbackClassName` (and updated the Map call site) with a JSDoc note that it only applies in the error state.

File: apps/geolibre-desktop/src/hooks/usePlugins.ts [hunk 8]
```diff
@@ -158,7 +175,16 @@ export function usePluginRegistry() {
     getProjectState: () => manager.getProjectState(),
     toggle: (id: string, appApi: ReturnType<typeof createAppAPI>) => {
       const before = JSON.stringify(projectPluginStateSnapshot());
-      manager.toggle(id, appApi);
+      // Plugin controls are imperative MapLibre code, so a throw here escapes
+      // React's error boundaries. Contain it so one bad plugin can't break the
+      // toggle handler — surface it in diagnostics instead. Return without
+      // persisting so a half-applied failure is not written to the project.
+      try {
+        manager.toggle(id, appApi);
```
[the remark this answers is not in the document]
Comment (giswqs) [reply apps/geolibre-desktop/src/hooks/usePlugins.ts:183]: Done in a25fa00 — added a comment documenting the known in-memory partial-mutation limitation (project persistence protected; in-memory not rolled back), as suggested.

File: tests/error-boundary.test.ts [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+  ));
+  ({ reportBoundaryError } = await import(
+    "../apps/geolibre-desktop/src/components/common/error-boundaries"
+  ));
+  ({ clearDiagnostics, getDiagnosticsSnapshot } = await import(
+    "../apps/geolibre-desktop/src/lib/diagnostics"
+  ));
+});
+
+describe("ErrorBoundary derived state", () => {
```
[the remark this answers is not in the document]
Comment (giswqs) [reply tests/error-boundary.test.ts:45]: Done in a25fa00 — added componentDidUpdate recovery-wiring tests (reset-on-change, no-reset-when-unchanged, no-reset-without-error). Used a reset() spy rather than asserting on state, since setState is a no-op on an unmounted instance.

File: packages/ui/src/components/error-boundary.tsx [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+ * lifecycle methods, and constructors of descendants. They do not catch errors
+ * in event handlers, asynchronous code, or imperative (non-React) code.
+ */
+export class ErrorBoundary extends React.Component<
+  ErrorBoundaryProps,
+  ErrorBoundaryState
+> {
+  state: ErrorBoundaryState = { error: null };
+
+  static getDerivedStateFromError(error: Error): ErrorBoundaryState {
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/ui/src/components/error-boundary.tsx:50]: Done in a25fa00 — added the JSDoc note that undefined<->non-empty transitions trigger a reset, while absent and empty are treated as equivalent.
